drivers/eeprom/eeprom_fm24w256.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2021 Caspar Friedrich <c.s.w.friedrich@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <devicetree.h>
+#include <devicetree/gpio.h>
+#include <drivers/eeprom.h>
+#include <drivers/gpio.h>
+#include <drivers/i2c.h>
+#include <logging/log.h>
+
+#define DT_DRV_COMPAT cypress_fm24w256
+
+// 256-Kbit ferroelectric random access memory logically organized as 32K × 8 bit
+#define SIZE KB(32)
+
+// Get the upper seven address bits and mask them properly
+#define OFFSET_HI(offset) ((offset >> 8) & BIT_MASK(7))
+
+// Get the lower eight address bits and mask them properly. Redundacy for consistency.
+#define OFFSET_LO(offset) ((offset >> 0) & BIT_MASK(8))
+
+LOG_MODULE_REGISTER(fm24w256, CONFIG_EEPROM_LOG_LEVEL);
+
+struct config {
+	struct i2c_dt_spec i2c_spec;
+	struct gpio_dt_spec wp_spec;
+	bool read_only;
+};
+
+static int configure_pin(const struct gpio_dt_spec *spec)
+{
+	if (!spec->port) {
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(spec->port)) {
+		return -ENODEV;
+	}
+
+	return gpio_pin_configure_dt(spec, GPIO_OUTPUT_INACTIVE);
+}
+
+static int init(const struct device *dev)
+{
+	const struct config *config = dev->config;
+	int ret;
+
+	/*
+	 * Write protect via GPIO is optional. Only configure if set.
+	 */
+	if (config->wp_spec.port) {
+		ret = configure_pin(&config->wp_spec);
+		if (ret) {
+			LOG_ERR("Unable to configure write protect pin: %d", ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int memory_operation(const struct device *dev, off_t offset, const void *buf, size_t buf_len,
+			    bool is_write)
+{
+	uint8_t addr_buf[] = {
+		OFFSET_HI(offset),
+		OFFSET_LO(offset),
+	};
+
+	struct i2c_msg msg[] = {
+		{
+			.buf = addr_buf,
+			.len = ARRAY_SIZE(addr_buf),
+			.flags = I2C_MSG_WRITE,
+		},
+		{
+			.buf = (uint8_t *)buf,
+			.len = buf_len,
+			.flags = (is_write ? I2C_MSG_WRITE : I2C_MSG_RESTART | I2C_MSG_READ) |
+				 I2C_MSG_STOP,
+		},
+	};
+
+	const struct config *config = dev->config;
+
+	if (is_write && config->read_only) {
+		LOG_ERR("Device is configured read only");
+		return -EACCES;
+	}
+
+	if (offset < 0 || offset > SIZE) {
+		LOG_ERR("Offset out of range: %d", offset);
+		return -EINVAL;
+	}
+
+	if (offset + buf_len > SIZE) {
+		LOG_ERR("Memory roll over");
+		return -EINVAL;
+	}
+
+	return i2c_transfer_dt(&config->i2c_spec, msg, ARRAY_SIZE(msg));
+}
+
+static int read(const struct device *dev, off_t offset, void *buf, size_t buf_len)
+{
+	LOG_DBG("About to read %u bytes from 0x%04x", buf_len, offset);
+	return memory_operation(dev, offset, buf, buf_len, false);
+}
+
+static int write(const struct device *dev, off_t offset, const void *buf, size_t buf_len)
```

Comment:
Just picking a random function to comment on here: It would be nice to use some more unique/descriptive names for the static functions. This greatly aids in debugging sessions, I think.

The common pattern in Zephyr drivers is to use the driver name as the first part of the function name (e.g. `fm24w256_write()`.

Reply:
done